### PR TITLE
AP_Follow: support for Mount following the lead vehicle in follow mode

### PR DESCRIPTION
**Pull request by khanasif786**
While In follow mode, Mount also follows the same sysid the vehicle is following.
Tested in SITL.

### CODE REVIEW

File: ArduCopter/mode_follow.cpp [hunk 1]
```diff
@@ -32,6 +32,17 @@ void ModeFollow::exit()
 
 void ModeFollow::run()
 {
+#if HAL_MOUNT_ENABLED
+    Location target_loc;
+    Vector3f veh_vel;
+    AP_Follow *f = AP_Follow::get_singleton();
```
Comment (rmackay9) [inline ArduCopter/mode_follow.cpp:38]: Txs for this.  It's looking pretty good, We should add a nullptr check of f and s.

File: ArduCopter/mode_follow.cpp [hunk 2]
```diff
@@ -32,6 +32,17 @@ void ModeFollow::exit()
...
+    AP_Mount *s = AP_Mount::get_singleton();
+    f->get_target_location_and_velocity(target_loc,veh_vel);
+    if(f->mount_follow_enabled()){
```
Comment (rmackay9) [inline ArduCopter/mode_follow.cpp:41]: nitpick: put a space between the if and the first bracket.

File: ArduCopter/mode_follow.cpp (hunk 1 above)
Comment (rmackay9) [reply ArduCopter/mode_follow.cpp:38]: actually it looks like the rest of ModeFollow::run() uses "g2.follow" so let's use that instead.

**Review by peterbarker — CHANGES_REQUESTED** [review]
This doesn't look right.  Mount can already follow a system ID - the only thing you might want to do is to add the option for follow mode to set the system ID for it to follow.

**Comment by rmackay9** [comment]
.. and one last little thing is that PRs should always have a description even if the title makes it pretty obvious.  .. normally that description should also include what testing has been done.

File: libraries/AP_Follow/AP_Follow.cpp [hunk 3]
```diff
@@ -130,6 +130,13 @@ const AP_Param::GroupInfo AP_Follow::var_info[] = {
     AP_GROUPINFO("_ALT_TYPE", 10, AP_Follow, _alt_type, AP_FOLLOW_ALT_TYPE_DEFAULT),
 #endif
 
+    // @Param: _OPTIONS
+    // @DisplayName: Follow options
+    // @Description: Follow options bitmask. Specifies mount behaviour in follow mode
+    // @Values: 0:None,1: Mount Follows lead vehicle
+    // @User: Standard
+    AP_GROUPINFO("_OPTIONS", 11, AP_Follow, _options, 0),
```
Comment (peterbarker) [inline libraries/AP_Follow/AP_Follow.cpp:138]: Can't use index 0 for parameters.  The maths doesn't work out in our fancy parameter storage system.

File: libraries/AP_Follow/AP_Follow.cpp [hunk 4]
```diff
@@ -130,6 +130,13 @@ const AP_Param::GroupInfo AP_Follow::var_info[] = {
     AP_GROUPINFO("_ALT_TYPE", 10, AP_Follow, _alt_type, AP_FOLLOW_ALT_TYPE_DEFAULT),
 #endif
 
+    // @Param: _OPTIONS
+    // @DisplayName: Follow options
+    // @Description: Follow options bitmask. Specifies mount behaviour in follow mode
```
Comment (peterbarker) [inline libraries/AP_Follow/AP_Follow.cpp:135]:
```suggestion
    // @Description: Follow options bitmask.
```

may be used for other things later

File: libraries/AP_Follow/AP_Follow.cpp [hunk 5]
```diff
@@ -130,6 +130,13 @@ const AP_Param::GroupInfo AP_Follow::var_info[] = {
...
+    // @Values: 0:None,1: Mount Follows lead vehicle
```
Comment (peterbarker) [inline libraries/AP_Follow/AP_Follow.cpp:136]:
```suggestion
    // @Values: 0:None,1: Mount Follows lead vehicle on mode enter
```

File: libraries/AP_Follow/AP_Follow.h [hunk 6]
```diff
@@ -28,6 +28,11 @@ class AP_Follow
 
 public:
 
+    // enum for FOLLOW_OPTIONS parameter
+    enum FollowOptions {
```
Comment (peterbarker) [inline libraries/AP_Follow/AP_Follow.h:32]:
```suggestion
    enum class Option {
```

singular and move redundant namespacing

File: libraries/AP_Follow/AP_Follow.h [hunk 7]
```diff
@@ -28,6 +28,11 @@ class AP_Follow
...
+        MOUNT_FOLLOW_ENABLED = 1
```
Comment (peterbarker) [inline libraries/AP_Follow/AP_Follow.h:33]:
```suggestion
        MOUNT_FOLLOW_ON_ENTER = 1
```

File: libraries/AP_Follow/AP_Follow.h [hunk 8]
```diff
@@ -98,6 +106,9 @@ class AP_Follow
     // get system time of last position update
     uint32_t get_last_update_ms() const { return _last_location_update_ms; }
 
+    // check if follow for mount is enabled
+    bool mount_follow_enabled() const { return (_options.get() & FollowOptions::MOUNT_FOLLOW_ENABLED) != 0; }
```
Comment (peterbarker) [inline libraries/AP_Follow/AP_Follow.h:110]:
```suggestion
    // returns true if a follow option enabled
    bool option_is_enabled(FollowOption option) const { return (_options.get() & (uint16_t)option) != 0; }
```

File: ArduCopter/mode_follow.cpp [hunk 9]
```diff
@@ -20,6 +20,15 @@ bool ModeFollow::init(const bool ignore_checks)
         gcs().send_text(MAV_SEVERITY_WARNING, "Set FOLL_ENABLE = 1");
         return false;
     }
+
+#if HAL_MOUNT_ENABLED
+    AP_Mount *mount = AP_Mount::get_singleton();
+    // follow the lead vehicle using sysid
+    if(g2.follow.mount_follow_enabled() && mount != nullptr) {
```
Comment (peterbarker) [inline ArduCopter/mode_follow.cpp:27]:
```suggestion
    if (options_is_enabled(Option::MOUNT_FOLLOW_ON_ENTER && mount != nullptr) {
```

**Review by peterbarker — APPROVED** [review]
Much nicer!

File: libraries/AP_Follow/AP_Follow.cpp (hunk 3 above)
Comment (rmackay9) [reply libraries/AP_Follow/AP_Follow.cpp:138]: @peterbarker I think this is using index 11 so it's OK.  The default is zero..

**Review by rmackay9 — APPROVED** [review]

File: libraries/AP_Follow/AP_Follow.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_Follow/AP_Follow.cpp:138]: Yes, sorry, I completely misread the line

**Review by peterbarker — APPROVED** [review]

**Review by peterbarker — CHANGES_REQUESTED** [review]
Ah.  This PR needs splitting as it crosses libraries and ArduCopter directories.

Please use `./Tools/gittools/git-subsystems-split` to make this multiple commits

**Review by peterbarker — APPROVED** [review]

File: libraries/AP_Follow/AP_Follow.h [hunk 10]
```diff
@@ -128,6 +139,7 @@ class AP_Follow
     AP_Int8     _yaw_behave;        // following vehicle's yaw/heading behaviour (see YAW_BEHAVE enum)
     AP_Int8     _alt_type;          // altitude source for follow mode
     AC_P        _p_pos;             // position error P controller
+    AP_Int16    _options;           // options for mount behaviour follow mode
```
Comment (magicrub) [inline libraries/AP_Follow/AP_Follow.h:142]: This doesn't hold up the merge or anything... but I don't see any harm in making all _OPTIONS params 32bit... We should just make it a habit to do that in the future.

**Pull request merged**